Fix flightcontrol test and usage bugs

sipsma opened the pull request:
Fixes #2270 

@tonistiigi The possible bug involving the executor host's file turned out to not be an issue because in that code branch the `hostname` var was in practice always set to `defaultHostname`. So this primarily fixes the bug in `computeBlobchain` and a few other smaller possible ones.

tonistiigi commented on solver/jobs.go:
```diff
@@ -715,7 +715,7 @@ func (s *sharedOp) CacheMap(ctx context.Context, index int) (resp *cacheMapResp,
 		return nil, err
 	}
-	res, err := s.g.Do(ctx, "cachemap", func(ctx context.Context) (ret interface{}, retErr error) {
+	res, err := s.g.Do(ctx, fmt.Sprintf("cachemap-%d", index), func(ctx context.Context) (ret interface{}, retErr error) {
```
That looks pretty bad. I wonder if it might have caused some cache misses. We should pick this commit to v0.9/

tonistiigi commented on cache/blobs.go:
```diff
@@ -46,121 +47,112 @@ func (sr *immutableRef) computeBlobChain(ctx context.Context, createIfNeeded boo
-					if err := ensureCompression(ctx, sr, desc, compressionType, s); err != nil {
-						return nil, err
-					}
+		refInfo := sr.Info()
```
This doesn't look quite equivalent. Previously if differ was in progress, `!createIfNeeded` would still wait for completion. And `!forceCompression` should not call differ twice, even in mediatype is different.

sipsma replied:
> And !forceCompression should not call differ twice, even in mediatype is different

Fixed

> Previously if differ was in progress, !createIfNeeded would still wait for completion

Agree that used to be the behavior, but was it on purpose? It seems strange to me for the behavior of `createIfNeeded` to be dependent on the value of `createIfNeeded` in other builds that happen to be running at the same time. If a user sets `createIfNeeded` to false, I feel like it makes the most sense to just never block on the differ running, especially since it's related to timing and thus never a deterministic thing anyways.

Let me know if I'm missing something, I can make it behave in the old way but it will convolute the code a bit more.

sipsma commented:
I added a unit test for GetRemote to the commit w/ changes to computeBlobChain (apparently didn't have coverage before).

After pushing the PR I see it keeps failing in CI only on Windows. It doesn't appear related to my change at all, it seems to just be because it's hitting a codepath previously never hit (it gets `"invalid windows mount type: 'bind'"` when trying to extract a blob to snapshotter).

So I updated it to just skip that test on Windows. Let me know if that's alright for now.

sipsma commented:
@tonistiigi let me know what you think of the updates

tonistiigi commented on cache/blobs.go:
```diff
@@ -44,29 +45,19 @@ func (sr *immutableRef) computeBlobChain(ctx context.Context, createIfNeeded boo
-				}
-				return nil, nil
-			} else if !createIfNeeded {
+		v, err := g.Do(ctx, fmt.Sprintf("%s-%t", sr.ID(), createIfNeeded), func(ctx context.Context) (interface{}, error) {
```
doesn't this skip the second mediatype or am I missing something. The call to `ensureCompression` looks to be in this block, therefore never called for the second attempt.

sipsma replied:
The call to `ensureCompression` is outside of the flightcontrol call, so multiple calls for the same blob w/ different mediatypes should still each call `ensureCompression` with their mediatype.

tonistiigi reviewed (commented):
Please verify this also works well with https://github.com/moby/buildkit/pull/2246 , whichever gets merged first.

sipsma commented:
> Please verify this also works well with #2246 , whichever gets merged first.

Yep, at first glance it looks conceptually compatible even if the merge conflicts will be slightly ugly. I also left a note on that other PR to update the new `TestGetRemote` added here in case this is merged first, which should help catch any regressions.

tonistiigi commented:
I think for v0.9 let's take only the cachemap (and maybe pull) fix.
